### PR TITLE
Update the reconnection sample to use the retry policy fully, rather than wrap another retry policy around method calls

### PR DESCRIPTION
**Pull request by drwill-ms**

### CODE REVIEW

File: iothub/device/samples/how to guides/DeviceReconnectionSample/CustomRetryPolicy.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            retryInterval = TimeSpan.Zero;
+            _logger.LogInformation($"Retry requested #{currentRetryCount} exception [{lastException.GetType()}: {lastException.Message}].");
+
+            if (currentRetryCount > MaxRetryCount)
+            {
+                _logger.LogError($"Retry requested #{currentRetryCount} of {MaxRetryCount} so giving up.");
+                return false;
+            }
+
+            if (lastException is IotHubClientException iotHubException
```
Comment (drwill-ms) [inline iothub/device/samples/how to guides/DeviceReconnectionSample/CustomRetryPolicy.cs:42]: Made simpler for v2.

File: iothub/device/samples/how to guides/DeviceReconnectionSample/DeviceReconnectionSample.cs [hunk 2]
```diff
@@ -61,50 +61,51 @@ public DeviceReconnectionSample(List<string> deviceConnectionStrings, Parameters
             _deviceConnectionStrings = deviceConnectionStrings;
             _logger.LogInformation($"Supplied with {_deviceConnectionStrings.Count} connection string(s).");
 
-            _clientOptions = new(parameters.GetHubTransportSettings()) { SdkAssignsMessageId = SdkAssignsMessageId.WhenUnset };
+            _clientOptions = new(parameters.GetHubTransportSettings())
+            {
```
Comment (drwill-ms) [inline iothub/device/samples/how to guides/DeviceReconnectionSample/DeviceReconnectionSample.cs:65]: prep for Nate's change to pass retry policy here.

File: iothub/device/samples/how to guides/DeviceReconnectionSample/DeviceReconnectionSample.cs [hunk 3]
```diff
@@ -166,142 +149,68 @@ private async Task InitializeAndSetupClientAsync(CancellationToken cancellationT
         // initialize or dispose the device client instance; the async method GetTwinAndDetectChangesAsync is implemented similarly for the same purpose.
         private async void ConnectionStatusChangeHandlerAsync(ConnectionStatusInfo connectionInfo)
         {
-            _logger.LogDebug($"Connection status changed: status={connectionInfo.Status}, reason={connectionInfo.ChangeReason}");
-            s_connectionStatus = connectionInfo.Status;
+            ConnectionStatus status = connectionInfo.Status;
+            ConnectionStatusChangeReason reason = connectionInfo.ChangeReason;
+            Console.WriteLine($"Connection status changed: status={status}, reason={reason}, recommendation={connectionInfo.RecommendedAction}");
+
+            // In our case, we can operate with more than 1 shared access key and attempt to fall back to a secondary.
```
Comment (drwill-ms) [inline iothub/device/samples/how to guides/DeviceReconnectionSample/DeviceReconnectionSample.cs:156]: Restoring this logic from the before time.

File: iothub/device/samples/how to guides/DeviceReconnectionSample/DeviceReconnectionSample.cs [hunk 4]
```diff
@@ -387,18 +285,15 @@ private static OutgoingMessage PrepareMessage(int messageId)
...
 
-        // If the client reports Connected status, it is already in operational state.
-        // If the client reports Disconnected_retrying status, it is trying to recover its connection.
-        // If the client reports Disconnected status, you will need to dispose and recreate the client.
-        // If the client reports Disabled status, you will need to dispose and recreate the client.
-        private bool ShouldClientBeInitialized(ConnectionStatus connectionStatus)
+        private bool ShouldClientBeInitialized()
         {
-            return (connectionStatus == ConnectionStatus.Disconnected || connectionStatus == ConnectionStatus.Closed)
+            return s_deviceClient?.ConnectionStatusInfo == null
```
Comment (drwill-ms) [inline iothub/device/samples/how to guides/DeviceReconnectionSample/DeviceReconnectionSample.cs:295]: Use the recommendation.

**Comment by drwill-ms** [comment]
Tests passed fine, and merge only affected a line that wouldn't impact tests. Build was validated locally.

**Review by abhipsaMisra — COMMENTED** [review]
looks good

**Pull request merged**